src/init.cpp

```diff
@@ -1246,10 +1246,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
         // Read asmap file if configured
         std::vector<bool> asmap;
         if (args.IsArgSet("-asmap")) {
-            fs::path asmap_path = fs::PathFromString(args.GetArg("-asmap", ""));
-            if (asmap_path.empty()) {
-                asmap_path = fs::PathFromString(DEFAULT_ASMAP_FILENAME);
-            }
+            fs::path asmap_path = args.GetPathArg("-asmap", DEFAULT_ASMAP_FILENAME);
```

Comment:
> I'm concerned this may not be equivalent on `WIN32`: `PathFromString` is not called on the default value in `GetPathArg`, so `DEFAULT_ASMAP_FILENAME`, a `char *`, is interpreted / converted by `GetPathArg` into a `fs::path`. I see `path(const char* c) : std::filesystem::path(c) {}` in `fs.h`. `PathFromString` is different on `WIN32`: `u8path(string)`.

That's correct. The behavior is inherently different on different platforms because windows path objects represent paths with `wchar_t` while POSIX paths are represented with `char`. On windows, the `fs::path` `const char*` constructor will call the `std::filesystem::path` constructor and do a locale conversion from `char` to `wchar_t` using the the current code page. Normally we want to avoid conversions based on the code page, and the reason we define an `fs::path` class instead of using `std::filesystem::path` is to prevent these conversions from happening accidentally. But in this case, DEFAULT_ASMAP_FILENAME is just ASCII text, so the conversion should be harmless and always result in the same path, regardless of code page. That's why the `const char*` constructor is allowed, and the reason for the [code comment](https://github.com/bitcoin/bitcoin/blob/848b11615b67a3c49f76ebbcaa241a322d8014d8/src/fs.h#L39) "Allow literal string arguments, which are safe as long as the literals are ASCII."

Reply:
re: https://github.com/bitcoin/bitcoin/pull/24306#discussion_r817419499

> We simply assume that char pointers are ASCII, but this is not true. In fact the unit tests get this wrong by passing char pointer to a string containing non ASCII chars (emojis).

I don't think I understand. I'm not aware of any places where unit tests or other code are doing something wrong. It is true that the `fs::path` interface does not guarantee there are no invalid conversions. It's just trying to add some safety and prevent unintended conversions. In any case, I'm happy to make changes if there are any suggestions.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/24306#discussion_r818566106

> I think the second argument is a char pointer, but it is not ascii

You're right, that's a real bug. Created #24469 to fix and maybe implement a followup